api/util/concurrent.go

```diff
@@ -0,0 +1,22 @@
+package util
+
+import "sync"
+
+// ConcurrentSlice is a slice that can be appended concurrently
+type ConcurrentSlice struct {
+	sync.RWMutex
+	items []interface{}
+}
+
+// Append safely appends items to a slice
+func (cs *ConcurrentSlice) Append(item interface{}) {
+	cs.Lock()
+	defer cs.Unlock()
+
+	cs.items = append(cs.items, item)
+}
+
+// GetItems gets the items from the list, should only be used when there are no operations working on it.
```

Comment:
> should only be used when there are no operations working on it
If so, then shall you enforce it with the mutex?

Reply:
A bit conflicted on this since the mutex requires to be enforced on every operation, like reading one element at a time. Even then there could be multiple operations could have been done on the slice itself. Since returning the slice is actually just returning the pointer to the backend "array" itself. A mutex will never protect this unless we return a deep copy of the slice, which in my opinion, is not required.